give_kudos.py

```diff
@@ -38,52 +37,67 @@ def email_login(self):
+        self.own_profile_id = self.page.locator("#athlete-profile .card-body > a").get_attribute('href').split("/athletes/")[1]
+        print(f"own_profile_id: {self.own_profile_id}")
+
 
-    def locate_kudos_buttons_and_maybe_give_kudos(self, button_locator) -> int:
+    def locate_kudos_buttons_and_maybe_give_kudos(self, web_feed_entry_locator) -> int:
         """
         input: playwright.locator class
         Returns count of kudos given.
         """
-        b_count= button_locator.count()
+        w_count = web_feed_entry_locator.count()
         given_count = 0
-        for i in range(b_count):
-            button = button_locator.nth(i)
-            if button.get_by_test_id("unfilled_kudos").count():
-                try:
-                    button.click(timeout=2000)
-                    given_count += 1
-                    time.sleep(1)
-                except TimeoutError:
-                    print("Click timed out.")
+        print(f"web feeds found: {w_count}")
+        for i in range(w_count):
+            # run condition check
+            curr_duration = time.time() - self.start_time
+            if curr_duration > self.max_run_duration:
+                print("Max run duration reached.")
+                break
+
+            web_feed = web_feed_entry_locator.nth(i)
+            p_count = web_feed.get_by_test_id("entry-header").count()
+
+            # check if activity has multiple participants
+            if p_count > 1:
+                print(f"Found multiple list entries: {p_count}")
+                for j in range(p_count):
+                    participant = web_feed.get_by_test_id("entry-header").nth(j)
+                    # ignore own activities
+                    if participant.get_by_test_id("owners-name").get_attribute('href').split("/athletes/")[1] != self.own_profile_id:
+                        kudos_container = web_feed.get_by_test_id("kudos_comments_container").nth(j)
+                        button = kudos_container.get_by_test_id("unfilled_kudos")
+                        if button.count() == 1:
+                            print("Feed has no kudo")
+                            button.click(timeout=0, no_wait_after=True)
+                            print("Kudos button clicked")
+                            given_count += 1
+                            time.sleep(1)
+            else:
+                # ignore own activities
+                if web_feed.get_by_test_id("owners-name").get_attribute('href').split("/athletes/")[1] != self.own_profile_id:
+                    button = web_feed.get_by_test_id("unfilled_kudos")
+                    if button.count() == 1:
+                        print("Feed has no kudo")
+                        button.click(timeout=0, no_wait_after=True)
+                        print("clicked")
+                        given_count += 1
+                        time.sleep(1)

```

Comment:
Repeated code that can be refactored into its own function based on DRY principle.

Reply:
done